core/src/main/java/com/linecorp/armeria/server/composition/AbstractCompositeService.java

```diff
@@ -165,8 +165,12 @@ public O serve(ServiceRequestContext ctx, I req) throws Exception {
 
             final ServiceRequestContext newCtx = new CompositeServiceRequestContext(
                     ctx, newRoute, result.routingResult().path());
-            try (SafeCloseable ignored = newCtx.push(false)) {
+            final ServiceRequestContext oldCtx = RequestContextThreadLocal.getAndSet(newCtx);
+            assert oldCtx == ctx;
+            try {
                 return result.value().serve(newCtx, req);
+            } finally {
+                RequestContextThreadLocal.set(oldCtx);
             }
```

Comment:
I guess we forgot to consider this case of replacing an existing root context with another one. Should we provide another method dedicated to this scenario?

Reply:
Added! 😄